packages/imt-rs/src/lib.rs

```diff
@@ -0,0 +1,320 @@
+use std::error::Error;
+
+use tiny_keccak::{Hasher, Keccak};
+
+pub struct IMT {
+    nodes: Vec<Vec<Option<IMTNode>>>,
+    zeroes: Vec<IMTNode>,
+    hash: IMTHashFunction,
+    depth: usize,
+    arity: usize,
+}
+
+struct IMTMerkleProof {
+    root: IMTNode,
+    leaf: IMTNode,
+    path_indices: Vec<usize>,
+    siblings: Vec<Vec<IMTNode>>,
+    leaf_index: usize,
+}
+
+
+impl IMT {
+    pub fn new(hash: IMTHashFunction, depth: usize, zero_value: IMTNode, arity: usize, leaves: Option<Vec<IMTNode>>) -> Result<IMT, &'static str> {
```

Comment:
we can take `Vec<IMTNode>>` instead of `Option<Vec<_>>`, cause we simply can pass empty vec, if we don't have any leaves to insert during initialization